### DISCUSSION TITLE
go.mod custom versioning regex results in no new digest though new version is correct

### DISCUSSION THREAD

**Discussion by trietsch** [renovatebot/renovate#25920] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

I'm trying to update a `go.mod` dependency, from a custom Go Modules Repository. The version is structured as follows:
`v<semver>-<build_timestamp>-<not_a_git_commit_digest>.<revision>` (e.g. `v1.28.1-20231026084439-2903466d71eb.1`)

What I'm trying to achieve, is to update the semver and revision, but keep the `build_timestamp` and `not_a_git_commit_digest` the same (e.g. `20231026084439-2903466d71eb` should remain the same).

I have the following configuration for the `gomod` manager:
```
  packageRules: [
{
      matchManagers: [
        "gomod"
      ],
      matchPackagePatterns: [
        "*"
      ],
      matchUpdateTypes: [
        "minor",
        "patch"
      ],
      groupName: "golang dependencies",
      groupSlug: "non-major-golang",
      automerge: true,
      platformAutomerge: true,
      addLabels: [
        "minor-patch",
        "go"
      ]
    },
    {
      matchManagers: [
        "gomod"
      ],
      matchPackagePatterns: [
        "*"
      ],
      matchUpdateTypes: [
        "major"
      ],
      groupName: "golang major dependencies",
      groupSlug: "golang",
      automerge: false,
      semanticCommitType: "feat",
      addLabels: [
        "major",
        "go"
      ]
    },
    {
      matchManagers: [
        "gomod"
      ],
      matchPackagePrefixes: [
        "buf.build/gen/go"
      ],
      digest: false,
      versioning: "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-(?<compatibility>.*)\\.(?<build>\\d+)$"
    }
  ]
```

When doing a dry run, I see the following passing by:

```
DEBUG: Could not determine new digest for update. (repository=getstrm/cli)
       "packageName": "buf.build/gen/go/getstrm/pace/protocolbuffers/go",
       "currentValue": "v1.28.1-20231121141154-e945a7aece2d.4",
       "datasource": "go",
       "newValue": "v1.31.0-20231121141154-e945a7aece2d.2",
       "bucket": "non-major"
```

Which indicates that Renovate is able to successfully determine the next version. I don't really understand the `Could not determine new digest` part. Looking at the source code of Renovate, I do see that it tries to determine the [new digest from the fetched releases](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/process/lookup/index.ts#L471).

What am I doing wrong here? 

Repositories are open source:
- [Shared config repo](https://github.com/getstrm/renovate-config/blob/main/base-renovate-config.json5)
- [Repo using Renovate](https://github.com/getstrm/cli/blob/main/renovate.json5)

### Logs (if relevant)

<details><summary>Logs</summary>

```
<omitted>
"gomod": [
           {
             "deps": [
               {
                 "managerData": {"lineNumber": 2},
                 "depName": "go",
                 "depType": "golang",
                 "currentValue": "1.20",
                 "datasource": "golang-version",
                 "versioning": "go-mod-directive",
                 "updates": [],
                 "packageName": "go",
                 "warnings": [],
                 "sourceUrl": "https://github.com/golang/go",
                 "registryUrl": "https://raw.githubusercontent.com/golang/website",
                 "homepage": "https://go.dev/",
                 "currentVersion": "1.21.4"
               },
               {
                 "managerData": {"lineNumber": 5, "multiLine": true},
                 "depName": "buf.build/gen/go/getstrm/pace/grpc/go",
                 "depType": "require",
                 "currentValue": "v1.3.0-20231121141154-e945a7aece2d.2",
                 "datasource": "go",
                 "currentDigest": "e945a7aece2d",
                 "digestOneAndOnly": true,
                 "updates": [],
                 "packageName": "buf.build/gen/go/getstrm/pace/grpc/go",
                 "versioning": "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-(?<compatibility>.*)\\.(?<build>\\d+)$",
                 "warnings": [
                   {
                     "message": "Could not determine new digest for update (go package buf.build/gen/go/getstrm/pace/grpc/go)",
                     "topic": "buf.build/gen/go/getstrm/pace/grpc/go"
                   }
                 ],
                 "currentVersion": "v1.3.0-20231121141154-e945a7aece2d.2",
                 "fixedVersion": "v1.3.0-20231121141154-e945a7aece2d.2"
               },
               {
                 "managerData": {"lineNumber": 6, "multiLine": true},
                 "depName": "buf.build/gen/go/getstrm/pace/protocolbuffers/go",
                 "depType": "require",
                 "currentValue": "v1.28.1-20231121141154-e945a7aece2d.4",
                 "datasource": "go",
                 "currentDigest": "e945a7aece2d",
                 "digestOneAndOnly": true,
                 "updates": [],
                 "packageName": "buf.build/gen/go/getstrm/pace/protocolbuffers/go",
                 "versioning": "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-(?<compatibility>.*)\\.(?<build>\\d+)$",
                 "warnings": [
                   {
                     "message": "Could not determine new digest for update (go package buf.build/gen/go/getstrm/pace/protocolbuffers/go)",
                     "topic": "buf.build/gen/go/getstrm/pace/protocolbuffers/go"
                   }
                 ],
                 "currentVersion": "v1.28.1-20231121141154-e945a7aece2d.4",
                 "isSingleVersion": true,
                 "fixedVersion": "v1.28.1-20231121141154-e945a7aece2d.4"
               }
<omitted>

```

</details>

**Comment by trietsch** [comment]
I did some research and discovered the following:
- The `currentDigest` is determined from the [value extracted](https://github.com/renovatebot/renovate/blob/3228c12e703d115a67ac9f7aee0194dfb019eb46/lib/modules/manager/gomod/extract.ts#L31) from the `go.mod`, based on [this regex pattern](https://github.com/renovatebot/renovate/blob/3228c12e703d115a67ac9f7aee0194dfb019eb46/lib/modules/datasource/go/index.ts#L36).
   For `v1.28.1-20231109150944-6af097c4673f.1`, the digest will be `6af097c4673f`. As said in my question above, this is undesirable, as `20231109150944-6af097c4673f` denotes the version of the Protobuf files used, and `1.28.1-1` denotes the version used for the Protobuf code generation.
- I'm unsure how the `newDigest` is determined. I have seen that this happens [here](https://github.com/renovatebot/renovate/blob/3228c12e703d115a67ac9f7aee0194dfb019eb46/lib/workers/repository/process/lookup/index.ts#L471), which for a Go datasource boils down to [this switch case statement](https://github.com/renovatebot/renovate/blob/3228c12e703d115a67ac9f7aee0194dfb019eb46/lib/modules/datasource/go/index.ts#L78), though from there on, I can't determine how it should be determined. My hunch is that it ends up at the [default null value](https://github.com/renovatebot/renovate/blob/3228c12e703d115a67ac9f7aee0194dfb019eb46/lib/modules/datasource/go/index.ts#L93), as I'm using `buf.build` for managing Protobuf Go dependencies.
- I've tried using a Regex manager with a `currentValue` named capturing group extraction that does not contain the `v` in the prefix as a workaround to avoid the `currentDigest` being set, but I had no success there.

Am I running into a bug? Or is it incorrect of `buf.build` to assume that a "commit hash pattern" (note my previous statement that this is not a git commit) can be used as a Proto version reference in the Go package version?

**Reply by trietsch** [reply]
Sorry to tag you @rarkins @viceice, but who maintains the Go data source and could help out here?

**Reply by rarkins** [reply]
There is no maintainer of the Go datasource or any other, so please do not tag us again in future. If I can help you, I already would. If you can produce a minimal reproduction then it increases the chances that someone can help you. So far what you describe above and show in your config is complicated, and likely unnecessarily so for the purposes of getting help

**Reply by trietsch** [reply]
Ok, apologies for tagging. Is it listed somewhere which data sources are officially maintained?

**Reply by rarkins** [reply]
What I meant by "there is no maintainer of the go datasource or any other" is that there's nobody whose duty is to answer you about go or any other datasource. This is open source and there are no SLAs. If we are not interested in your problem, and we don't think it affects a wide number of users, then we can choose to ignore it. If you propose an elegant solution, we likely will accept it. But we are not duty bound to look deeply into problems which affect only one or a small subset of users.

**Reply by trietsch** [reply]
My question was with all the respect and sincerity, just trying to see if this might be interesting for you. Responding the way you do is understandable and I know that developing OS software is often with people that don't know the unwritten rules, including all the effort and workload that comes with OS. 
So sorry for bothering you, but I do have to say that it's not nice reading a message with such a sentiment, when my message was not demanding any help, nor expecting any work or effort. I may be the exception, but not everyone's intentions are as you expect.

I'll close the question now.

**Reply by rarkins** [reply]
Apologies if I misinterpreted your intentions. I was getting "I insist to speak with the ~~manager~~maintainer" vibes. Sometimes people have a hard time accepting that in Open Source you may not get answers to all questions, and nobody owes you anwers.

I would like to help you, it's just that I need to prioritize it against many other questions which affect more than one user. Are you able to simulate this problem with a public repo and public dependency? i.e. a public, minimal reproduction.

I'm guessing that what you require is:
- Stop Renovate from attempting to look up "digest" updates
- Use `versionCompatibility` or `allowedVersions` to enforce same timestamp/digest values but allow updating of the version/revision part

**Reply by trietsch** [reply]
No worries, glad that you respond the way you do. And I fully get that at some point you get frustrated with people demanding stuff. You're not working for charity either and you have the hard task of prioritizing. I was creating a MCVE, but then I noticed that your other comment below regarding a related PR might solve my issue, which after some tweaking of my config, it does indeed.

Thanks for the pointer and for Renovate!

**Comment by rarkins** [comment, marked as answer]
May be relevant: https://github.com/renovatebot/renovate/pull/26746

**Reply by trietsch** [reply]
The issue for my repo is fixed, by instructing the `gomod` package manager to treat `buf.build` [remote packages differently](https://github.com/getstrm/cli/pull/42) when considering the versioning. The bottom line is, to ensure that the `20231026084439-2903466d71eb` is a the `compatibility` part of the version, and therefore should not be updated (this resembles the version of your protocol buffer files, therefore you should not want to auto update these).

Important remarks:
* only set the `versioning` to strategy regex. Also setting the `extractVersion` to the same regex pattern results in a `Could not determine new digest for update`
* Use Renovate Bot version > `37.140.13`, as this includes a fix for setting the digest version for extracted Go datasource versions.

The following should be a complete example of upgrading buf.build remote go packages:
```json5
{
  packageRules: [
    {
      matchManagers: [
        "gomod"
      ],
      matchPackagePatterns: [
        "*"
      ],
      matchUpdateTypes: [
        "minor",
        "patch"
      ],
      groupName: "golang dependencies",
      groupSlug: "non-major-golang",
      automerge: true,
      platformAutomerge: true,
      addLabels: [
        "minor-patch",
        "go"
      ]
    },
    {
      matchManagers: [
        "gomod"
      ],
      matchPackagePatterns: [
        "*"
      ],
      matchUpdateTypes: [
        "major"
      ],
      groupName: "golang major dependencies",
      groupSlug: "golang",
      automerge: false,
      semanticCommitType: "feat",
      addLabels: [
        "major",
        "go"
      ]
    },
    {
      matchManagers: [
        "gomod"
      ],
      matchPackagePrefixes: [
        "buf.build/gen/go/getstrm/pace"
      ],
      // Ensure that the Protos version is treated as a compatibility version part. For example:
      // 20240129161937-d5ced9c85c25 in the dependency version v1.3.0-20240129161937-d5ced9c85c25.2
      // should not be updated
      versioning: "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-(?<compatibility>.*\\.)(?<build>\\d+)$"
    }
  ]
}
```

**Reply by trietsch** [reply]
[Here's an example PR](https://github.com/getstrm/cli/pull/43) that maintains the Buf Protos version correctly.
